### ISSUE TITLE
Add notifications to ApplicationState

### ISSUE THREAD

**Issue by dferber90** [commercetools/merchant-center-application-kit#68]
Now that we have the beautiful `ApplicationStateProvider` which serves as the-one-true-intersection-point between applications and and the AppShell, I think it might be a good idea to enhance it even more.

One of the things which is currently still inconvenient is showing notifications in the MC. I think we could move this to the `ApplicationStateProvider` as well.

Right now dispatching notifications works like this

```jsx
import { connect } from 'react-redux';
import * as globalActions from '@commercetools-frontend/actions-global';

// your component
const SomeComponent = props => (
  <button
    onClick={() =>
      props.showNotification({
        kind: "success",
        domain: DOMAINS.SIDE,
        text: "hello"
      })
    }
  >
    Show Notification
  </button>
);

export default connect(
  null,
  { showNotification: globalActions.showNotification }
)(SomeComponent);
```

However this could get as simple as

```jsx
const SomeComponent = props => (
  <GetApplicationState
    render={applicationState => (
      <button
        onClick={() =>
          applicationState.showNotification({
            kind: "success",
            domain: DOMAINS.SIDE,
            text: props.intl.formatMessage(messages.lineItemUpdated)
          })
        }
      >
        Show Notification
      </button>
    )}
  />
);
```

This would allow us to also test notifications dispatched by components more easily.

```jsx
import SomeComponent from './some-component'
import { render, fireEvent } from '@commercetools-frontend/application-kit/test-utils'

it("should dispatch a notification when clicked", () => {
  const { getByText, notifications } = render(<SomeComponent />);
  fireEvent.click(getByText("Show Notification"));
  expect(notifications).toContainEqual({
    kind: "success",
    domain: DOMAINS.SIDE,
    text: "hello"
  });
});
```

**Comment by dferber90** [comment]
One thing which is kinda weird is that the `applicationState` now isn't just *state*  anymore, but it evolves to more than that. It is now the intersection of the application and the application shell. It provides the API to exchange information between the two. It is no longer just a getter, but also a dispatcher/setter.

Given the new capabilities it is weird to call `applicationState.showNotification`, because a state is usually something which gets read only. It is also a bit weird that the component enabling this is called `GetApplicationState` when we are not really getting a state, but we only want to use it to emit an event (show a notification).

Should we maybe rename `applicationState` and `GetApplicationState`?

I can see more usage-as-setters coming in the future:
- setting the `dataLocale`?
- dealing with *view settings* (not sure about this one)?

**Comment by emmenko** [comment]
Thanks, that's definitely a good idea to look into.

The main question I have at the moment is about **how much stuff** are we willing to put into the context.
I think the naming "application state" is still valid though. Also `notifications` are part of the state.

What we can do though, is maybe separate the options that are passed to the consumers:

```js
// now
<GetApplicationState render={applicationState => ()} />

// we can change it to pass named params
<GetApplicationState render={({ applicationState, showNotification, ... }) => ()} />
```

So `applicationState` becomes a named param, which allows us to pass as many other params as we want.

**Comment by tdeekens** [comment]
It's a nice idea taken away the burden of showing notifications, I agree. But it feels somewhat like feeding a moloch to me. The name then would be surely off as it's not only state then. 

Renaming would be apt but not sure I sure I am fine without then re-defining the overall responsibility of the module. I guess it would be something like:

```js
<GetApplicationContext render={({state, dispatch}) => ()} />
```

while `dispatch` can have `dispatch.showNotification`. I am really a bit torn.

**Comment by emmenko** [comment]
If we need to rename/restructure something, this is the time to do so before the release. Otherwise it would be another breaking change...

**Comment by emmenko** [comment]
PS: "application context" I suppose is also fine and understandable 👌

**Comment by tdeekens** [comment]
> PS: "application context" I suppose is also fine and understandable 👌

Yeah it's a bit loaded term for me in React but conveys what we mean maybe. Given that I am just not 100% sure what the 2nd arg would be named:

1. `dispatch`
   - Kind of good as it indicates that something changes
   - Somewhat loaded and giving the assumption it's always redux actions (which it won't have to be)
2. `mutators`
   - Kind of vague if you just show a notification but it's what happens after all
3. `setters`
   - Not sure

**Comment by emmenko** [comment]
We don't necessarily need to have only 3-4 top level keys, I think we can also put "as many as we want".

```js
{
  state,
  showNotification,
  setProjectDataLocale,
  // ...
}
```

**Comment by emmenko** [comment]
So, if we want to we can then rename the component and pass the state as a named param.

Then as follow ups we can iterate on "adding more features" to it.

WDYT?

**Comment by montezume** [comment]
Works for me. I like ApplicationContext

**Comment by tdeekens** [comment]
I am fine either way: grouped arguments or not.

**Comment by emmenko** [comment]
Alright, I'll spin up a PR then

**Comment by dferber90** [comment]
Sorry, I just realised that I screwed up the code of the original issue. I now edited the second snippet to represent my idea more clearly.

Another example of something which we could provide using *"`GetApplicationState`"* (or whichever name we end up using) came to mind: tracking. We have got an imperative tracking function at the moment, which would be nice to also provide through the context.

---

> Alright, I'll spin up a PR then

Lets hold our horses before rushing in one direction 🐴 To me it's not entirely clear yet what the big picture is.

---

It seems like this unearthed a deeper architectural question: *"How do we want applications to interact with app-shell?"*. I don't have a good answer to it yet, but I'll share my understanding of the decision we'll need to take below.

> Renaming would be apt but not sure I sure I am fine without then re-defining the overall responsibility of the module

I like this. It will align us on where we're planning to go with the overall app-shell architecture.

At the moment, the AppShell is responsible for fetching the initial data and for setting up a bunch of providers.

To me the responsibility of the module is to be the intersection between the AppShell and the application itself. The AppShell sets a bunch of things up which developers of applications then don't need to worry about: user, project, permissions, feature flags, routing, tracking, intl, data locale, data fetching.

The application itself then uses the "consumer" components of these:
- user, project, permissions, data locale: `GetApplicationState`
- feature flags: `BranchOnFeatureToggle`, .. (everything `flopflip` provides)
- routing: `Route`, `Switch`, .. (everything `react-router` provides)
- tracking: `injectTracking` (from `core/components/tracking`) [not yet]
- intl: `FormattedMessage`, `injectIntl` (everything `react-intl` provides)
- data fetching: `Query` (everything `apollo` provides)
- data fetching: the http client for the `node-sdk`


I like that we're embracing the libraries `AppShell` is built on instead of trying to wrap everything (so no moloch).

It also seems like there are concerns (`notifications`, `tracking`, `dataLocale`) which we've built custom solutions for. The question is now whether we want to keep using the custom solutions as standalone tools or to unify them in `ApplicationContext`.

Do we want `AppliationContext` to be the go-to thing to access tracking, notifications, user, project, permissions, data locale or do we want to use more fine-grained consumers like  `GetApplicationState`, `GetNotifications`, `GetTracking` (hypothetical).

From a testing-perspective it doesn't really matter which road we take as they would all still get set up in `AppShell`, so we can mock them anyways (we would probably need to move notifications out of redux though).

To me the discussion shouldn't yet be about the actual API we're going to build a provider with, but we should first settle on the architecture we envision for applications.

Let's continue in person?
